src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal sealed class OpenSslPkcs12Reader : IDisposable
+    {
+        private readonly SafePkcs12Handle _pkcs12Handle;
+        private SafeEvpPkeyHandle _evpPkeyHandle;
+        private SafeX509Handle _x509Handle;
+        private SafeX509StackHandle _caStackHandle;
+
+        private OpenSslPkcs12Reader(SafePkcs12Handle pkcs12Handle)
+        {
+            _pkcs12Handle = pkcs12Handle;
+        }
+
+        public unsafe static OpenSslPkcs12Reader TryRead(byte[] data)
+        {
+            SafePkcs12Handle handle = Interop.libcrypto.OpenSslD2I(
+                (ptr, b, i) => Interop.libcrypto.d2i_PKCS12(ptr, b, i),
+                data,
+                checkHandle: false);
+
+            if (handle != null && !handle.IsInvalid)
+            {
+                return new OpenSslPkcs12Reader(handle);
+            }
+
+            return null;
+        }
+
+        public static OpenSslPkcs12Reader TryRead(SafeBioHandle fileBio)
```

Comment:
Yeah, I went back and forth on this one.  I liked the way the current approach captures the disposable object and doesn't leak it into the outer scope; but I disliked that I named it Try without being bool-returning and out-value-emitting; but Try did convey the semantic of "this might not work, but it won't throw, so pay attention".

I'll just change it to
```C#
OpenSslPkcs12Reader reader;
if (TryRead(handle, out reader))
{
    using (reader)
    {
        ...
    }
}
```

and if the variable scope bothers me when I do it, there's always an extra curly brace pair.